// @markbind/core/src/Page/PageConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PageConfig = void 0;
/**
 * A page configuration object used to construct a {@link Page}.
 * Its properties will never be modified by {@link Page} itself.
 */
class PageConfig {
    constructor(args) {
        this.asset = args.asset;
        this.baseUrlMap = args.baseUrlMap;
        this.dev = args.dev;
        this.faviconUrl = args.faviconUrl;
        this.frontmatterOverride = args.frontmatterOverride || {};
        this.layout = args.layout;
        this.layoutsAssetPath = args.layoutsAssetPath;
        this.pluginManager = args.pluginManager;
        this.resultPath = args.resultPath;
        this.rootPath = args.rootPath;
        this.searchable = args.searchable;
        this.siteLinkManager = args.siteLinkManager;
        this.siteOutputPath = args.siteOutputPath;
        this.sourcePath = args.sourcePath;
        this.src = args.src;
        this.title = args.title;
        this.template = args.template;
        this.variableProcessor = args.variableProcessor;
        this.addressablePagesSource = args.addressablePagesSource;
        this.layoutManager = args.layoutManager;
    }
}
exports.PageConfig = PageConfig;
