// @markbind/core/src/Site/constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._ = exports.LAZY_LOADING_SITE_FILE_NAME = exports.SITE_CONFIG_NAME = exports.INDEX_MARKDOWN_FILE = void 0;
const difference_1 = __importDefault(require("lodash/difference"));
const differenceWith_1 = __importDefault(require("lodash/differenceWith"));
const flatMap_1 = __importDefault(require("lodash/flatMap"));
const has_1 = __importDefault(require("lodash/has"));
const isBoolean_1 = __importDefault(require("lodash/isBoolean"));
const isEmpty_1 = __importDefault(require("lodash/isEmpty"));
const isEqual_1 = __importDefault(require("lodash/isEqual"));
const isUndefined_1 = __importDefault(require("lodash/isUndefined"));
const noop_1 = __importDefault(require("lodash/noop"));
const omitBy_1 = __importDefault(require("lodash/omitBy"));
const startCase_1 = __importDefault(require("lodash/startCase"));
const union_1 = __importDefault(require("lodash/union"));
const uniq_1 = __importDefault(require("lodash/uniq"));
exports.INDEX_MARKDOWN_FILE = 'index.md';
exports.SITE_CONFIG_NAME = 'site.json';
exports.LAZY_LOADING_SITE_FILE_NAME = 'LazyLiveReloadLoadingSite.html';
exports._ = {
    difference: difference_1.default,
    differenceWith: differenceWith_1.default,
    flatMap: flatMap_1.default,
    has: has_1.default,
    isUndefined: isUndefined_1.default,
    isEqual: isEqual_1.default,
    isEmpty: isEmpty_1.default,
    isBoolean: isBoolean_1.default,
    noop: noop_1.default,
    omitBy: omitBy_1.default,
    startCase: startCase_1.default,
    union: union_1.default,
    uniq: uniq_1.default,
};
