// @markbind/core/src/html/Context.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Context = void 0;
const CyclicReferenceError_1 = __importDefault(require("../errors/CyclicReferenceError"));
class Context {
    constructor(cwf, 
    /**
     * Array of cwfs file paths processed so far
     */
    callStack, variables, processingOptions = {}) {
        this.cwf = cwf;
        this.callStack = callStack;
        this.variables = variables;
        this.processingOptions = processingOptions;
    }
    addCwfToCallstack(cwf) {
        this.callStack.push(cwf);
    }
    hasExceededMaxCallstackSize() {
        return this.callStack.length > CyclicReferenceError_1.default.MAX_RECURSIVE_DEPTH;
    }
    clone() {
        return new Context(this.cwf, this.callStack.map(cwf => cwf), this.variables, this.processingOptions);
    }
}
exports.Context = Context;
