// @markbind/core/src/plugins/mermaid.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
const cheerio_1 = __importDefault(require("cheerio"));
const DEFAULT_CDN_ADDRESS = 'https://unpkg.com/mermaid@10/dist/mermaid.esm.min.mjs';
function genScript(address) {
    return `<script type="module">
    import mermaid from '${address || DEFAULT_CDN_ADDRESS}';
    document.addEventListener('DOMContentLoaded', () => {
      mermaid.initialize({});
      Vue.directive('mermaid', {
        inserted: function(el) {
          mermaid.run({
            nodes: [el]
          });
        }
      });
    });
  </script>`;
}
module.exports = {
    tagConfig: {
        mermaid: {
            isSpecial: true,
        },
    },
    getScripts: (pluginContext) => [genScript(pluginContext.address)],
    postRender: (pluginContext, frontmatter, content) => {
        const $ = cheerio_1.default.load(content);
        $('mermaid').each((index, node) => {
            const $node = $(node);
            $node.replaceWith(`<div class="mermaid" v-mermaid>${$node.html()}</div>`);
        });
        return $.html();
    },
};
